Fix #155016 for Docathon - convert rst to markdown

j-silv opened the pull request:
Used [rst2myst tool](https://rst-to-myst.readthedocs.io/en/latest/)

One note is that "Created On" and "Last Updated On" banner doesn't show in the markdown files... I'm not sure if that's just an artifact of my local build though.

Fixes #155016

Docs comparison (check out the 'new' whenever docs build)

1. cuda ([old](https://docs.pytorch.org/docs/main/cuda.html) vs. [new](https://docs-preview.pytorch.org/pytorch/pytorch/155198/cuda.html)) 
2. cuda.tunable ([old](https://docs.pytorch.org/docs/main/cuda.tunable.html) vs. [new](https://docs-preview.pytorch.org/pytorch/pytorch/155198/cuda.tunable.html))
3. leave cudnn_persistent_rnn.rst as is because it's reused in docstrings
4. cudnn_rnn_determinism.rst as is because it's reused in docstrings.
5. data ([old](https://docs.pytorch.org/docs/main/data.html) vs. [new](https://docs-preview.pytorch.org/pytorch/pytorch/155198/data.html))

cc @svekars @sekyondaMeta @AlannaBurke

j-silv commented:
@pytorchbot labels docathon-h1-2025 module: docs topic: not user facing topic: docs

svekars commented on docs/source/cudnn_persistent_rnn.md:
```diff
@@ -0,0 +1,9 @@
+:::{note}
+If the following conditions are satisfied:
+1\) cudnn is enabled,
+2\) input data is on the GPU
+3\) input data has dtype `torch.float16`
+4\) V100 GPU is used,
+5\) input data is not in `PackedSequence` format
+persistent algorithm can be selected to improve performance.
+:::
```
```suggestion
```{admonition}
If the following conditions are satisfied:
1) cudnn is enabled,
2) input data is on the GPU
3) input data has dtype `torch.float16`
4) V100 GPU is used,
5) input data is not in `PackedSequence` format
persistent algorithm can be selected to improve performance.
```
```

j-silv replied:
I kept the `note`, as it looks identical to the original PR (`admonition` didn't show for some reason) but I made sure the actual bullet points showed.

svekars commented on docs/source/data.md:
```diff
@@ -0,0 +1,519 @@
+constructor is {attr}`dataset`, which indicates a dataset object to load data
+from. PyTorch supports two different types of datasets:
+
+- [map-style datasets],
```
Does the link work? should if be: ` [map-style datasets](Map-style-datasets)` ?

svekars commented on docs/source/data.md:
````diff
@@ -0,0 +1,519 @@
+
+In this case, loading from a map-style dataset is roughly equivalent with:
+
+```
````
```suggestion
```python
```

svekars commented on docs/source/data.md:
````diff
@@ -0,0 +1,519 @@
+
+and loading from an iterable-style dataset is roughly equivalent with:
+
+```
````
```suggestion
```python
```

svekars commented on docs/source/data.md:
````diff
@@ -0,0 +1,519 @@
+
+Example:
+
+```
````
```suggestion
```python
```

svekars reviewed (commented):
Looking good overall - can you resolve the merge conflict?

j-silv commented:
Ah I see. I was rebasing on `viable/strict` but it looks like `cuda.rst` was touched in the main branch in the meantime.

Concerning your changes (i.e. adding `python` as the code block language), I think I could fork the rst2myst tool and add that functionality fairly easy. For now though, I'll just add the language keyword manually.

Also, I think the fact that the "Created On" and "Last Updated On" banner don't show is actually a limitation of MySt, or there is some a needed config option somewhere. 

Finally, it seems that the `rst2myst` is really bad at handling anchor links. 

For example, it converted this in `data.rst`:

```
* `single- and multi-process data loading <Single- and Multi-process Data Loading_>`_,

* `automatic memory pinning <Memory Pinning_>`_.
```

To this:

```
- [single- and multi-process data loading],
- [automatic memory pinning][memory pinning].
```

These are definitely not valid anchor links. 

I saw that in `torch.md`, the author is using `refs` and in my case we would want:

```
- {ref}`map-style and iterable-style datasets <dataset-types>`,
- {ref}`customizing data loading order <data-loading-order-and-sampler>`,
- {ref}`automatic batching <loading-batched-and-non-batched-data>`,
```

and then wherever that header is, we would put:

```
(dataset-types)=
## Dataset Types

...etc
```

Just want to confirm that this is the style we should adopt. I tried to use explicit headers as [specified per here](https://myst-parser.readthedocs.io/en/latest/syntax/cross-referencing.html#creating-explicit-targets), but those did not output correctly when running `make html`.

```
(heading-target)=
### Heading

[reference1](#heading-target)
```

j-silv committed:
Fix #155016 for Docathon - convert rst to markdown

Used rst2myst tool: https://rst-to-myst.readthedocs.io/en/latest/
for the initial conversion and then hand-modified the broken links and
python syntax highlighting in the code blocks.

Squash svekars comments in the original PR.

Also this commit is broken into 2 so that we can keep the git log history of
data.md.

svekars commented:
I think we need to fix [these errors](https://github.com/pytorch/pytorch/actions/runs/15498885987/job/43643374890?pr=155198) by fixing rst -> .md extensions in the include directives in these locations:

- https://github.com/pytorch/pytorch/blob/main/torch/nn/modules/rnn.py#L589
- https://github.com/pytorch/pytorch/blob/main/torch/nn/modules/rnn.py#L946
- https://github.com/pytorch/pytorch/blob/main/torch/nn/modules/rnn.py#L591
- https://github.com/pytorch/pytorch/blob/main/torch/nn/modules/rnn.py#L948
- https://github.com/pytorch/pytorch/blob/main/torch/ao/nn/quantized/dynamic/modules/rnn.py#L797

albanD reviewed (approved):
python code change sounds good!

svekars commented:
I think this won't work unfortunately since we probably can't use .md includes in docstrings themselves. Sorry for the confusion. We should undo the changes in the rnn.py files, and keep cudnn_rnn_determinism.rst, cudnn_persistent_rnn.rst as .rst files.

svekars commented:
It looks like[ it doesn't like ](https://github.com/pytorch/pytorch/actions/runs/15564323233/job/43825993847?pr=155198#step:11:827)the reference type of link in the docstring itself. You can change the ref to use a direct link instead:
`multiprocessing context <https://docs.python.org/3/library/multiprocessing.html#contexts-and-start-methods>`_ in [here](https://github.com/pytorch/pytorch/blob/c88e87f3553b7ef0b743fb030317fabb6fe2dc29/torch/utils/data/dataloader.py#L178) and delete: https://github.com/pytorch/pytorch/blob/c88e87f3553b7ef0b743fb030317fabb6fe2dc29/torch/utils/data/dataloader.py#L227-L228

j-silv commented:
Ahhh now the docstring is too [wide](https://github.com/pytorch/pytorch/actions/runs/15566788198/job/43832778680#step:15:86):

```
  Advice (FLAKE8) B950
    line too long (174 > 120 characters)
    See https://github.com/PyCQA/flake8-bugbear#list-of-warnings

         175  |            worker subprocess with the worker id (an int in ``[0, num_workers - 1]``) as
         176  |            input, after seeding and before data loading. (default: ``None``)
         177  |        multiprocessing_context (str or multiprocessing.context.BaseContext, optional): If
    >>>  178  |            ``None``, the default `multiprocessing context <https://docs.python.org/3/library/multiprocessing.html#contexts-and-start-methods>`_ of your operating system will
```
Any way around this?

svekars commented on torch/utils/data/dataloader.py:
```diff
@@ -175,7 +175,7 @@ class DataLoader(Generic[_T_co]):
             input, after seeding and before data loading. (default: ``None``)
         multiprocessing_context (str or multiprocessing.context.BaseContext, optional): If
-            ``None``, the default `multiprocessing context`_ of your operating system will
+            ``None``, the default `multiprocessing context <https://docs.python.org/3/library/multiprocessing.html#contexts-and-start-methods>`_ of your operating system will
```
```suggestion
            ``None``, the default `multiprocessing context
            <https://docs.python.org/3/library/multiprocessing.html#contexts-and-start-methods>`_ 
            of your operating system will
```

svekars replied:
try breaking into multiple lines.

svekars commented on torch/utils/data/dataloader.py:
```diff
@@ -175,7 +175,9 @@ class DataLoader(Generic[_T_co]):
         multiprocessing_context (str or multiprocessing.context.BaseContext, optional): If
-            ``None``, the default `multiprocessing context`_ of your operating system will
+            ``None``, the default 
+            `multiprocessing context <https://docs.python.org/3/library/multiprocessing.html#contexts-and-start-methods>`_ # noqa D401
```
```suggestion
            `multiprocessing context <https://docs.python.org/3/library/multiprocessing.html#contexts-and-start-methods>`_ # noqa: D401
```

j-silv commented:
One sec - `data.md` has no git log history. I'll break this into another commit to keep it

j-silv committed:
Fix #155016 - modify data.md (second commit)

Second commit whereas the first was just a git mv to
maintain history of data.rst

svekars commented:
you can just undo your changes to data.rst on this branch (save them somewhere), then:
1. From docs dir: git checkout origin/main -- source/optim.rst - delete .md I guess as well.
2. git mv source/data.rst source/data.md
3. Make a commit: git commit -m "Renamed source/data.rst to source/data.md".
5. Can check git log: git log -- follow -- data.md - this should show you a bunch of commits.
6. Push into this PR the above commit.
7. Copy paste changes you've saved in step 1 into data.md. Commit/push into this PR.

j-silv commented:
Yep should be good now.
